### ISSUE TITLE
False positive when merging dict types defined recursively

### ISSUE THREAD

**Issue by fakuivan** [microsoft/pyright#4881]
**Describe the bug**
When merging dict types defined recursively, key types of a specific type will not typecheck correctly agains a union that contains it.

**To Reproduce**
```python
# A tree where nodes can be strings or ints
KeyStrOrInt = dict[str | int, "KeyStrOrInt"]
# A tree where the first nodes are strings
FirstLevelKeyStr = dict[str, KeyStrOrInt]

def test(targs: FirstLevelKeyStr, **kwargs: FirstLevelKeyStr) -> KeyStrOrInt:
    # kwargs is a tree where the first and second nodes are strings
    # We should be able to merge these.
    merged = targs | kwargs
    return merged
```

```text
(variable) merged: dict[str, KeyStrOrInt | FirstLevelKeyStr]

Expression of type "dict[str, KeyStrOrInt | FirstLevelKeyStr]" cannot be assigned to return type "KeyStrOrInt"
  "dict[str, KeyStrOrInt | FirstLevelKeyStr]" is incompatible with "KeyStrOrInt"
    TypeVar "_KT@dict" is invariant
      Type "str" cannot be assigned to type "str | int"
        "str" is incompatible with "int"
```

This line in particular I believe is false: `Type "str" cannot be assigned to type "str | int"`
**Expected behavior**
Typechecks correctly

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
def foo(self) -> str:
    return 3
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Pyright VS code extension: v1.1.301

**Additional context**

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The resulting type from the `|` operation is `dict[str, dict[str | int, KeyStrOrInt] | dict[str, dict[str | int, KeyStrOrInt]]` which is not compatible with `dict[str | int, KeyStrOrInt]`. The type parameters for `dict` are invariant, so they must match exactly.

Mypy generates the same error in this case.

Using recursive type aliases in the presence of invariant type parameters requires special care. One workaround in this case is to change the first line of `test` to:
```python
merged = KeyStrOrInt(**targs) | KeyStrOrInt(**kwargs)
```

**Comment by fakuivan** [comment]
Why are type parameters invariant in this case? It seems like they are variant for this case:

```python
def test2(a: dict[int | str, int], b: dict[int, int]) -> dict[int | str, int]:
    return a | b
```

How does introducing recursion change this?

**Comment by erictraut** [comment]
The `dict` class defines two type parameters, and both of them are invariant. 

Here's why it's important for its type parameters to be invariant.
```python
KeyStrOrInt = dict[str | int, "KeyStrOrInt"]
FirstLevelKeyStr = dict[str, KeyStrOrInt]

def test(d: FirstLevelKeyStr):
    # The following is a type violation, but let's assume
    # for a moment that it's not.
    x: KeyStrOrInt = d
    x[0] = {}

y: FirstLevelKeyStr = {}
test(y)

print(y) # (0: {}), which isn't valid for FirstLevelKeyStr
```

**Comment by erictraut** [comment]
By default, all class-scoped TypeVars are invariant unless they are explicitly created as covariant or contravariant. You may find PEP 483 and 484 useful reading.
